Attached the Color Sensor Plugin

stevendes opened the pull request:
![robot_sensor](https://user-images.githubusercontent.com/53347092/67509227-e6c1c100-f668-11e9-9ab9-9083f7e798d1.png)


**UPDATE 1**

- Attached the previously created Color Sensor plugin to the Create model

**HOW IT WORKS**

It responds by sending an Bool message to the topic related to the specified camera, for example, the left sensor publishes in the topic "left_yellow_sensor".

**ISSUES**

- Only works when a there is a subscription in the "image_raw" topic of every camera.

**UPDATE 2**

- Addressed all issues and comments, working fine.

stevendes committed:
Add color sensor

Applied changes requested

Adressed the issues commented

Fixed type

Update sensor model

Added camera mesh

Add color sensor

Applied changes requested

Adressed the issues commented

Fixed type

Add color sensor

Applied changes requested

Adressed the issues commented

Fixed type

Add color sensor

Applied changes requested

Adressed the issues commented

Fixed type

Add color sensor

Applied changes requested

Adressed the issues commented

Fixed type

changed xacro files

Modifying the original plugin to detect RGB

Added the plugin in the two color sensors from previous PR

Adding the files for the sensor plugin

Add action messages

Addressed some comments and added realistic mesh

Addressed requested changes, added comments

Fixed publication in sensor plugin

Update camera plugin

Addressed all the comments

stevendes commented:
@eborghi10 @ernestmc @serraramiro1 @LucasJSch

eborghi10 commented:
Some things @stevendes:

1. Was the issue with `alwaysOn` fixed?
2. `try2` is your target branch? Consider using better names.
3. You have conflicts.

eborghi10 commented on ca_description/urdf/color_sensor.xacro:
```diff
@@ -0,0 +1,95 @@
+<?xml version="1.0" ?>
+<robot xmlns:xacro="http://ros.org/wiki/xacro">
+
+<xacro:macro name="color_sensor" params="prefix parent color">
```
The origin block is missing as Xacro parameter.

eborghi10 commented on ca_description/urdf/color_sensor.xacro:
```diff
@@ -0,0 +1,95 @@
+
+<xacro:macro name="color_sensor" params="prefix parent color">
+ <joint name="${prefix}_color_joint" type="fixed">
+    <xacro:if value="${prefix == 'left'}">
```
This approach is too complex. You are passing `origin` incorrectly.

eborghi10 commented on ca_description/urdf/color_sensor.xacro:
```diff
@@ -0,0 +1,95 @@
+  </joint>
+
+  <link name="${prefix}_color_link">
+    <xacro:dummy_inertia/>
```
Is this needed? The joint is fixed.

eborghi10 commented on ca_description/urdf/color_sensor.xacro:
```diff
@@ -0,0 +1,95 @@
+
+  <link name="${prefix}_color_link">
+    <xacro:dummy_inertia/>
+    <origin xyz="0 0 0" rpy="0 0 0"/>
```
Is this needed? The parameter is the default one.

eborghi10 commented on ca_description/urdf/color_sensor.xacro:
```diff
@@ -0,0 +1,95 @@
+    <origin xyz="0 0 0" rpy="0 0 0"/>
+    <visual>
+        <geometry>
+  	       <box size="0.015 0.015 0.015"/>
```
Can't you use a realistic mesh?

eborghi10 commented on ca_description/urdf/color_sensor.xacro:
```diff
@@ -0,0 +1,95 @@
+  </joint>
+
+  <link name="${prefix}_color_optical_link">
+    <xacro:dummy_inertia/>
```
Necessary?

eborghi10 commented on ca_description/urdf/color_sensor.xacro:
```diff
@@ -0,0 +1,95 @@
+         <always_on>true</always_on>
+         <update_rate>60.0</update_rate>
+         <imageTopicName>${prefix}_rgb/image_raw</imageTopicName>
+         <depthImageTopicName>${prefix}_depth/image_raw</depthImageTopicName>
```
Are all of these parameters necessary?

eborghi10 commented on ca_description/urdf/color_sensor.xacro:
```diff
@@ -0,0 +1,95 @@
+          <gvalue_min>230</gvalue_min>
+          <bvalue_max>140</bvalue_max>
+          <bvalue_min>0</bvalue_min>
+          <PublisherName>${prefix}_${color}_sensor</PublisherName>
```
You're using the same parameter name independently of the color.

eborghi10 commented on ca_description/urdf/color_sensor.xacro:
```diff
@@ -0,0 +1,95 @@
+          <bvalue_min>230</bvalue_min>
+          <PublisherName>${prefix}_${color}_sensor</PublisherName>
+         </xacro:if>
+         <FactorColorperImage>0.08</FactorColorperImage>
```
`FactorColorPerImage`?

eborghi10 commented on ca_description/urdf/create_base.xacro:
```diff
@@ -44,6 +45,24 @@
+    <!-- Creating the constants to introduce origin -->
+
+    <xacro:property name="left_origin">
+      <origin xyz="${diameter/2 * 0.5 + 0.01} ${diameter/2 * 0.87 + 0.01} ${size_z}" rpy="0 ${pi/4} 0" />
```
This is not the correct way of passing Xacro block parameters. Again, [here](http://wiki.ros.org/xacro#Macros) is the link.

eborghi10 commented on ca_description/urdf/create_base_gazebo.xacro:
```diff
@@ -26,7 +26,7 @@
         <publishWheelJointState>true</publishWheelJointState>
         <legacyMode>false</legacyMode>
-        <odometrySource>world</odometrySource>
+        <odometrySource>map</odometrySource>
```
These changes are already in the original repo so you'll need to update your branches.

eborghi10 commented on ca_gazebo/CMakeLists.txt:
```diff
@@ -24,16 +24,25 @@ else()
-  include_directories(include
+set(CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS}")
+
+
```
Too many empty lines.

eborghi10 commented on ca_gazebo/CMakeLists.txt:
```diff
@@ -24,16 +24,25 @@ else()
 
-	link_directories(${GAZEBO_LIBRARY_DIRS})
-  include_directories(include
+set(CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS}")
```
Why did you add this here having [this one](https://github.com/stevendes/create_autonomy/blob/8481c5fd5120ed942570ff5e8b1de2d0c5dc97fe/ca_gazebo/CMakeLists.txt#L5)?

eborghi10 commented on ca_gazebo/CMakeLists.txt:
```diff
@@ -24,16 +24,25 @@ else()
+include_directories(include
     ${catkin_INCLUDE_DIRS}
-		${GAZEBO_INCLUDE_DIRS}
+	${GAZEBO_INCLUDE_DIRS}
```
There is a difference in tabulation.

eborghi10 commented on ca_gazebo/CMakeLists.txt:
```diff
@@ -24,16 +24,25 @@ else()
+	${Boost_INCLUDE_DIR}
   )
 
 	add_library(create_bumper_plugin src/create_bumper_plugin.cpp)
```
Organize the plugins like this:

```
add_library()
add_dependencies()
target_link_libraries()

add_library()
add_dependencies()
target_link_libraries()
```

eborghi10 commented on ca_gazebo/include/ca_gazebo/color_sensor_plugin.h:
```diff
@@ -0,0 +1,48 @@
+    {
+      class GazeboRosColor : public CameraPlugin, GazeboRosCameraUtils
+      {
+        /// \brief Constructor
```
Organize your class components like this:

```C++
class GazeboRosColor
{
public:
...
...
private:
...
...
}
```

stevendes commented:
> Some things @stevendes:
> 
>     1. Was the issue with `alwaysOn` fixed?
> 
>     2. `try2` is your target branch? Consider using better names.
> 
>     3. You have conflicts.

Hello @eborghi10 

1- The issue wasn't fixed yet, all the things that I tried didn't work out
2- Yes haha, you already pointed me that out when I used that branch for the PR, but because this is that same branch, has the same name since I didn't closed that PR and opened a new one, and I used better name to my branches since then
3- I would check it and then tag everyone again

eborghi10 commented on ca_gazebo/include/ca_gazebo/color_sensor_plugin.h:
```diff
@@ -0,0 +1,48 @@
+        protected: virtual void OnNewFrame(const unsigned char *_image,
+        unsigned int _width, unsigned int _height,
+        unsigned int _depth, const std::string &_format);
+        unsigned int _rvalue_max;
```
Add comments to explain better each variable/method.

eborghi10 commented on ca_gazebo/include/ca_gazebo/color_sensor_plugin.h:
```diff
@@ -0,0 +1,48 @@
+        unsigned int _bvalue_max;
+        unsigned int _bvalue_min; 
+        std::string _publisher_name;
+        unsigned int _factor_color_per_image;
```
"frame" not "image".

eborghi10 commented on ca_gazebo/include/ca_gazebo/color_sensor_plugin.h:
```diff
@@ -0,0 +1,48 @@
+        unsigned int _factor_color_per_image;
+
+        ros::NodeHandle _nh;
+        ros::Publisher _sensorPublisher;
```
Are you going to use camelCase or snake_case?

ernestmc replied:
Class data members in C++ are named differently than in Python. The underscore goes in the end. See here:
https://google.github.io/styleguide/cppguide.html#Variable_Names

eborghi10 commented on ca_gazebo/launch/create_empty_world.launch:
```diff
@@ -14,13 +14,23 @@
 		<arg name="world_name" value="$(find ca_gazebo)/worlds/$(arg env).world"/>
 	</include>
+	
+	<group if="$(eval str(arg('env'))=='empty')">
```
Is this a fix that is not in the original repo? If so, consider creating a PR with the fix.

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+  ////////////////////////////////////////////////////////////////////////////////
+  // Constructor
+  GazeboRosColor::GazeboRosColor():
+  _nh("color_sensor_plugin"),
```
Use 2 spaces of tabulation here.

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+
+    CameraPlugin::Load(_parent, _sdf);
+    // copying from CameraPlugin into GazeboRosCameraUtils
+    this->parentSensor_ = this->parentSensor;
```
Are all these parameters being used?

stevendes replied:
Yes, removing one of them makes the plugin misbehave

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+    this->depth_ = this->depth;
+    this->format_ = this->format;
+    this->camera_ = this->camera;
+    this->_rvalue_max = _sdf->GetElement("rvalue_max")->Get<double>();
```
This is not the way of getting the parameters. Use the templetized methods. For example:

```C++
this->_rvalue_max = _sdf->GetElement<double>("rvalue_max");
```

stevendes replied:
This gives me the following error:
```
/create_ws/src/create_autonomy/ca_gazebo/src/color_sensor_plugin.cpp:56:42: err$
r: expected primary-expression before ‘double’
     this->rvalue_max_ = _sdf->GetElement<double>("rvalue_max");
```
The info that I found says that Element was a class of Gazebo in the earlier versions, but I couldn't find how to use it in the 9.0 API, the old one however does follow the format that I was using:

[http://osrf-distributions.s3.amazonaws.com/gazebo/api/1.3.0/classsdf_1_1Element.html#a29fa74bb6bf29701e82b6353e732d31d](url)

stevendes replied:
Erasing the "Element" part apparently solves the issue

`this->rvalue_max_ = _sdf->Get<double>("rvalue_max");`

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+    this->_bvalue_max = _sdf->GetElement("bvalue_max")->Get<double>();
+    this->_bvalue_min = _sdf->GetElement("bvalue_min")->Get<double>();
+    this->_publisher_name = _sdf->Get<std::string>("PublisherName");
+     _sensorPublisher = _nh.advertise<std_msgs::Bool>(this->_publisher_name, 1);
```
Why is this not using `this`?

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+    unsigned int _width, unsigned int _height, unsigned int _depth,
+    const std::string &_format)
+  {
+    static int seq=0;
```
Why is this for?

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+
+    this->sensor_update_time_ = this->parentSensor_->LastUpdateTime();
+
+    if (!this->parentSensor->IsActive())
```
Since always have the camera enabled, is this necessary?

stevendes replied:
Addressed, however since the camera still isn't always enable I would leave the conversation open

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+    {
+      if ((*this->image_connect_count_) > 0)
+      {
+        common::Time cur_time = this->world_->SimTime();
```
Make it const.

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+      if ((*this->image_connect_count_) > 0)
+      {
+        common::Time cur_time = this->world_->SimTime();
+        if (cur_time - this->last_update_time_ >= this->update_period_)
```
This variable (`last_update_time_`) is never set before. You need to do it in `Load()`.

stevendes replied:
Putting `this->last_update_time_ = this->world_->SimTime();` in the CameraPlugin::Load breaks Gazebo

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+        common::Time cur_time = this->world_->SimTime();
+        if (cur_time - this->last_update_time_ >= this->update_period_)
+        {
+          this->PutCameraData(_image);
```
Is this necessary?

stevendes replied:
Gazebo misbehave without it

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+
+          for (int i=0; i<_width*_height*3 ;i=i+3)
+          {
+              R=_image[i] ;
```
Spaces between operators.

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+              }
+            }
+
+          if(_desired_color_in_pixel/_pixel_count > this->_factor_color_per_image)
```
```C++
_output = _desired_color_in_pixel/_pixel_count > this->_factor_color_per_image;
```

eborghi10 commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+          {
+            _output=false;
+          }
+        _sensorPublisher.publish(_output); 
```
Apply [this](http://wiki.ros.org/roscpp/Overview/Publishers%20and%20Subscribers#Intraprocess_Publishing) workaround.

eborghi10 commented:
> > 1. Was the issue with `alwaysOn` fixed?
> >  
> 1- The issue wasn't fixed yet, all the things that I tried didn't work out

@serraramiro1 (Tag Lucas too because I can't), could you check if you find out what's going on?

serraramiro1 commented:
> > > 1. Was the issue with `alwaysOn` fixed?
> > 
> > 
> > 1- The issue wasn't fixed yet, all the things that I tried didn't work out
> 
> @serraramiro1 (Tag Lucas too because I can't), could you check if you find out what's going on?

The way i fixed this, was adding this line to the load function.
```cpp
  this->parentSensor_->SetActive(true);
```
Didn't find any documentation for that though

ernestmc commented on ca_gazebo/include/ca_gazebo/color_sensor_plugin.h:
```diff
@@ -0,0 +1,48 @@
+      {
+        /// \brief Constructor
+        /// \param parent The parent entity, must be a Model or a Sensor
+        public: GazeboRosColor();
```
Where is the parameter?

ernestmc commented on ca_gazebo/include/ca_gazebo/color_sensor_plugin.h:
```diff
@@ -0,0 +1,48 @@
+        public: ~GazeboRosColor();
+    
+        /// \brief Load the plugin
+        /// \param take in SDF root element
```
Document each parameter separately

ernestmc commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,131 @@
+#include <gazebo/sensors/SensorTypes.hh>
+
+
+#include "std_msgs/Bool.h"
```
Order your includes. See here:
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

ernestmc reviewed (changes requested):
Left some comments

eborghi10 commented:
I have an error when I try to compile your branch:

```
In file included from /opt/ros/kinetic/include/ros/serialization.h:37:0,
                 from /opt/ros/kinetic/include/ros/publisher.h:34,
                 from /opt/ros/kinetic/include/ros/node_handle.h:32,
                 from /opt/ros/kinetic/include/ros/ros.h:45,
                 from /create_ws/src/create_autonomy/ca_gazebo/src/color_sensor_plugin.cpp:2:
/opt/ros/kinetic/include/ros/message_traits.h: In instantiation of ‘static const char* ros::message_traits::MD5Sum<M>::value(const M&) [with M = bool]’:
/opt/ros/kinetic/include/ros/message_traits.h:255:102:   required from ‘const char* ros::message_traits::md5sum(const M&) [with M = bool]’
/opt/ros/kinetic/include/ros/publisher.h:112:7:   required from ‘void ros::Publisher::publish(const M&) const [with M = bool]’
/create_ws/src/create_autonomy/ca_gazebo/src/color_sensor_plugin.cpp:123:41:   required from here
/opt/ros/kinetic/include/ros/message_traits.h:126:34: error: request for member ‘__getMD5Sum’ in ‘m’, which is of non-class type ‘const bool’
     return m.__getMD5Sum().c_str();
                                  ^
/opt/ros/kinetic/include/ros/message_traits.h: In instantiation of ‘static const char* ros::message_traits::DataType<M>::value(const M&) [with M = bool]’:
/opt/ros/kinetic/include/ros/message_traits.h:264:104:   required from ‘const char* ros::message_traits::datatype(const M&) [with M = bool]’
/opt/ros/kinetic/include/ros/publisher.h:112:7:   required from ‘void ros::Publisher::publish(const M&) const [with M = bool]’
/create_ws/src/create_autonomy/ca_gazebo/src/color_sensor_plugin.cpp:123:41:   required from here
/opt/ros/kinetic/include/ros/message_traits.h:143:36: error: request for member ‘__getDataType’ in ‘m’, which is of non-class type ‘const bool’
     return m.__getDataType().c_str();
                                    ^
create_autonomy/ca_gazebo/CMakeFiles/color_sensor_plugin.dir/build.make:62: recipe for target 'create_autonomy/ca_gazebo/CMakeFiles/color_sensor_plugin.dir/src/color_sensor_plugin.cpp.o' failed
make[2]: *** [create_autonomy/ca_gazebo/CMakeFiles/color_sensor_plugin.dir/src/color_sensor_plugin.cpp.o] Error 1
CMakeFiles/Makefile2:5478: recipe for target 'create_autonomy/ca_gazebo/CMakeFiles/color_sensor_plugin.dir/all' failed
make[1]: *** [create_autonomy/ca_gazebo/CMakeFiles/color_sensor_plugin.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
Invoking "make -j8 -l8" failed
```

@stevendes I found where the error is but it is up to you to solve this particular issue.
**Hint:** it's related to a ROS message.

eborghi10 commented:
Some things to improve here:

- You didn't address most of my comments.
- You have to improve a lot the tabulations and empty lines.

eborghi10 commented:
If I run this: `roslaunch ca_gazebo create_autorace.launch`, the simulation breaks:

```
XML parsing error: not well-formed (invalid token): line 101, column 1
when instantiating macro: create_base (/create_ws/src/create_autonomy/ca_description/urdf/create_base.xacro)in file: /create_ws/src/create_autonomy/ca_description/urdf/color_sensor.xacroincluded from: /create_ws/src/create_autonomy/ca_description/urdf/create_2.xacroCheck that: - Your XML is well-formed - You have the xacro xmlns declaration: xmlns:xacro="http://www.ros.org/wiki/xacro"while processing /create_ws/src/create_autonomy/ca_gazebo/launch/create_empty_world.launch:while processing /create_ws/src/create_autonomy/ca_gazebo/launch/spawn_multirobot.launch:while processing /create_ws/src/create_autonomy/ca_gazebo/launch/spawn_create.launch:while processing /create_ws/src/create_autonomy/ca_description/launch/create_description.launch:Invalid <param> tag: Cannot load command parameter [robot_description]: command [/opt/ros/kinetic/lib/xacro/xacro --inorder /create_ws/src/create_autonomy/ca_description/urdf/create_2.xacro visualize:=false] returned with code [2]. 

Param xml is <param command="$(find xacro)/xacro --inorder $(arg model) $(arg xacro_args)" name="robot_description"/>
The traceback for the exception was written to the log file
```

stevendes commented:
@eborghi10 I still didn't push the file with the addressed commits, the ones that are open hasn't been addressed yet, I don't have errors locally so maybe they are because the last commit is only to resolve the conflicts that were present, I would push it now and please let me know if the error is still present.

eborghi10 commented:
I wasn't able to test your code. What should I do? Which topics should I look at?

eborghi10 commented:
You have conflicts with the base branch.

stevendes commented:
@eborghi10 addressed most of the comments, the issue with the activation is still present but I'm moving forward with the actionlib task, then I would return to this and try to solve it again

LucasJSch commented on ca_gazebo/include/ca_gazebo/color_sensor_plugin.h:
```diff
@@ -0,0 +1,63 @@
+
+      };
+    }
+    #endif
```
https://google.github.io/styleguide/cppguide.html#The__define_Guard

LucasJSch commented on ca_gazebo/include/ca_gazebo/color_sensor_plugin.h:
```diff
@@ -0,0 +1,63 @@
+        ros::Publisher sensor_publisher_;
+
+      };
+    }
```
https://google.github.io/styleguide/cppguide.html#Namespaces

LucasJSch commented on ca_gazebo/include/ca_gazebo/color_sensor_plugin.h:
```diff
@@ -0,0 +1,63 @@
+        unsigned int gvalue_max_;
+        unsigned int gvalue_min_;
+        unsigned int bvalue_max_;
+        unsigned int bvalue_min_;
```
Are these completely necessary? I suppose it makes the code more parametrizable and thus more portable, but it has a trade-off that makes the code more difficult to understand because it is longer. So, reconsider using these parameters.

LucasJSch commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,121 @@
+#include "gazebo_plugins/gazebo_ros_camera.h"
+#include <gazebo/sensors/Sensor.hh>
+#include <gazebo/sensors/CameraSensor.hh>
+#include <gazebo/sensors/SensorTypes.hh>
```
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

LucasJSch commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,121 @@
+    else
+    {
+      if ((*this->image_connect_count_) > 0)
+      {*/
```
Delete if not necessary

LucasJSch commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,121 @@
+
+    int R,G,B;
+
+    double desired_color_in_pixel_=0, pixel_count_=0;
```
maybe you can name it `goal_color`. It's shorter :)

LucasJSch commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,121 @@
+      this->PublishCameraInfo();
+      this->last_update_time_ = cur_time;
+
+      bool output_;
```
Consider using a more representative name

eborghi10 replied:
Address this comment, please.

LucasJSch commented on ca_gazebo/src/color_sensor_plugin.cpp:
```diff
@@ -0,0 +1,121 @@
+
+      bool output_;
+
+      for (int i=0; i<_width*_height*3 ;i=i+3)
```
Spaces between operations (e.g. `i = 0; i < _width (...)`), and also add this way: `i += 3`

eborghi10 replied:
Address this comment and others too.

LucasJSch commented on ca_tools/action/FollowLine.action:
```diff
@@ -0,0 +1,10 @@
+
+---
+# Feedback
+int32 number_of_minutes
```
Newline at EOF

LucasJSch reviewed (commented):
Left some comments, PTAL.

eborghi10 commented on ca_gazebo/include/ca_gazebo/color_sensor_plugin.h:
```diff
@@ -0,0 +1,64 @@
+
+    namespace gazebo
+    {
+      class GazeboRosColor : public CameraPlugin, GazeboRosCameraUtils
```
Does this work if you're not subscribed to the camera topic?

eborghi10 replied:
Hint: it seems so, check [here](https://bitbucket.org/osrf/gazebo/src/default/plugins/CameraPlugin.cc). If not, you should inherit from `SensorPlugin`.

stevendes replied:
It works

eborghi10 commented on ca_gazebo/include/ca_gazebo/color_sensor_plugin.h:
```diff
@@ -0,0 +1,64 @@
+
+        private:
+
+        /*
```
You have to order the variables like this:

```
# comment1
variable1;
# comment2
variable2;
...
```

eborghi10 commented on ca_gazebo/launch/create_empty_world.launch:
```diff
@@ -30,7 +30,7 @@
   <!-- map_server -->
   <include file="$(find ca_move_base)/launch/map_server.launch">
-	<arg name="env" value="$(arg env)"/>
+	<arg name="env" value="empty"/>
```
Why is this change here?

eborghi10 reviewed (changes requested):
I made some comments, PTAL.

stevendes commented:
Addressed all the comments
